### PR TITLE
[RTM] Use ausi/slug-generator for the creation of aliases

### PR DESCRIPTION
**Pull request by ausi**
As discussed in #954

@contao/developers Is an event class the right thing for this use case?

ToDo:

- [x] Wait until CLDR 32.0 is released (19th October) and tag version 1.0.0 of ausi/slug-generator
- [x] Create pull requests for news-bundle and events-bundle: contao/news-bundle#20, contao/calendar-bundle#14

### CODE REVIEW

**Comment by agoat** [comment]
Have only looked quickly over it. Great.

Just as a note: Your package require `php: ^7.1`. If this will be merged, contao should also raise the minimum php version to 7.1.

**Comment by Toflar** [comment]
It already is ;) See the develop branch :)

**Comment by ausi** [comment]
Yes, the PHP requirement is the same as in Contao 4.5. But it also requires `ext-intl`, `ext-mbstring` and `ext-pcre`, maybe we should discuss if it is OK to require these extensions?

**Comment by leofeyer** [comment]
I guess so. Is there an alternative?

**Comment by Toflar** [comment]
Of course! All of them are required for a decent multilingual (and thus multibyte/utf-8) setup.

**Comment by ausi** [comment]
`ext-pcre` is already required by Contao and `patchwork/utf8`.

`ext-intl` and `ext-mbstring` are already required by `league/uri` which itself is required since Contao 4.1.

Therefore I think I can keep the requirements of the slug library as they are now.

**Comment by ausi** [comment]
The `de-ASCII` rules are included in the latest CLDR version 32-beta3, so I released `ausi/slug-generator` version 1.0 and updated this PR.

It should be RTM now 🎉

File: tests/Slug/ValidCharactersTest.php [hunk 1]
```diff
@@ -50,23 +54,43 @@ function (SlugValidCharactersEvent $event) use (&$path): bool {
...
-        ];
+        $translator
+            ->expects($this->at(0))
+            ->method('trans')
+            ->with('MSC.validCharacters.unicodeLowercase')
+            ->willReturn('Unicode numbers and lowercase letters')
+        ;
 
+        $translator
+            ->expects($this->at(1))
```
Comment (ausi) [inline tests/Slug/ValidCharactersTest.php]: Is it necessary to test how the tranlator is used?

File: src/Slug/ValidCharacters.php [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+    /**
+     * @var TranslatorInterface
+     */
+    private $translator;
+
+    /**
+     * @var string[]
+     */
+    private $defaultOptions = [
```
Comment (leofeyer) [inline src/Slug/ValidCharacters.php]: Could this also be a constant or static property?

File: src/Event/ContaoCoreEvents.php [hunk 3]
```diff
@@ -60,7 +60,7 @@
     public const BACKEND_MENU_BUILD = 'contao.backend_menu_build';
 
     /**
-     * The contao.slug_valid_characters event is triggered when the options for the valid characters setting for aliases are generated.
+     * The contao.slug_valid_characters event is triggered when the valid slug characters option is generated.
```
Comment (ausi) [inline src/Event/ContaoCoreEvents.php]: I think this should be plural “options are generated”

File: src/Resources/contao/languages/en/tl_page.xlf [hunk 4]
```diff
@@ -99,10 +99,10 @@
         <source>The date and time format string will be parsed with the PHP date() function.</source>
       </trans-unit>
       <trans-unit id="tl_page.validAliasCharacters.0">
-        <source>Valid characters for aliases</source>
+        <source>Valid alias characters</source>
       </trans-unit>
       <trans-unit id="tl_page.validAliasCharacters.1">
-        <source>When aliases get created automatically, this character set gets used for the conversion.</source>
+        <source>Please choose the character for auto-generating page aliases.</source>
```
Comment (ausi) [inline src/Resources/contao/languages/en/tl_page.xlf]: Also plural “characters”

File: src/Slug/ValidCharacters.php (hunk 2 above)
Comment (ausi) [reply src/Slug/ValidCharacters.php]: Yes, a private constant should be fine.

File: src/Event/ContaoCoreEvents.php (hunk 3 above)
Comment (leofeyer) [reply src/Event/ContaoCoreEvents.php]: Are there several valid slug characters options?

File: src/Resources/contao/languages/en/tl_page.xlf (hunk 4 above)
Comment (leofeyer) [reply src/Resources/contao/languages/en/tl_page.xlf]: True.

File: tests/Slug/ValidCharactersTest.php (hunk 1 above)
Comment (leofeyer) [reply tests/Slug/ValidCharactersTest.php]: Probably not. 😄

Comment (leofeyer) [reply tests/Slug/ValidCharactersTest.php]: Although the test assures that the class iterates over all four default options.

Comment (ausi) [reply tests/Slug/ValidCharactersTest.php]: But the order should not be relevant. I think we should remove the translator tests here.

File: src/Event/ContaoCoreEvents.php (hunk 3 above)
Comment (ausi) [reply src/Event/ContaoCoreEvents.php]: Yes, by default there are four options: unicode lowercase, unicode, ASCII lowercase and ASCII.

**Comment by leofeyer** [comment]
Very good job @ausi. Thanks a bunch.

**Comment by backbone87** [comment]
So 4.4 LTS does not and will not support this?

**Comment by ausi** [comment]
@backbone87 No, but there is an extension which backports the new behaviour to 4.4: [ausi/contao-slug-backport](https://github.com/ausi/contao-slug-backport)

**Comment by backbone87** [comment]
@ausi is it worth to add a contao specific impl of Ausi\SlugGenerator\SlugGenerator that provides a method with a PageModel argument, that automatically extracts the slug configuration and builds it? that would remove the duplicated code that gathers this configuration in various listeners and eases the integration for 3rd party bundles

**Comment by ausi** [comment]
I’m not sure if this is necessary. It’s not that much code that transforms the page model to the options array:

```php
$slugOptions['locale'] = $objPage->language;
if ($objPage->validAliasCharacters) {
    $slugOptions['validChars'] = $objPage->validAliasCharacters;
}
```

**Comment by backbone87** [comment]
well, you also must make sure, that $page->loadDetails() is executed and possibly strip insert tags.

further StringUtil::generateAlias should be deprecated and changed to use the slug generator

**Comment by ausi** [comment]
> further StringUtil::generateAlias should be deprecated and changed to use the slug generator

sounds reasonable, I think you should open a new ticket for that.

**Comment by ausi** [comment]
Related: #1334

**Pull request merged**